File: cmd/osqueryi-mcp/executor.go
package main

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"log/slog"
	"os/exec"
	"regexp"
	"sort"
	"strings"
	"sync"
	"time"
)

type Executor struct {
	binaryPath string
	timeout    time.Duration
	mu         sync.RWMutex
	tables     []string
	schemas    map[string][]TableColumn
}

func NewExecutor(binaryPath string, timeout time.Duration) *Executor {
	return &Executor{
		binaryPath: binaryPath,
		timeout:    timeout,
		schemas:    make(map[string][]TableColumn),
	}
}

type TableColumn struct {
	CID       string `json:"cid"`
	DfltValue string `json:"dflt_value"`
	Name      string `json:"name"`
	NotNull   string `json:"notnull"`
	PK        string `json:"pk"`
	Type      string `json:"type"`
}

type SearchMatch struct {
	TableName       string   `json:"table_name"`
	MatchReasons    []string `json:"match_reasons"`
	MatchingColumns []string `json:"matching_columns,omitempty"`
}

func (e *Executor) runSQL(ctx context.Context, sql string) ([]byte, error) {
	ctx, cancel := context.WithTimeout(ctx, e.timeout)
	defer cancel()

	slog.Debug("executing_sql", "sql", sql)
	cmd := exec.CommandContext(ctx, e.binaryPath, "--json", "--config_path=/dev/null", sql)
	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	start := time.Now()
	err := cmd.Run()
	duration := time.Since(start)

	if err != nil {
		errMsg := strings.TrimSpace(stderr.String())
		if errMsg == "" {
			slog.Debug("exec_failed", "sql", sql, "error", err, "duration_ms", duration.Milliseconds())
			return nil, err
		}
		slog.Debug("exec_failed", "sql", sql, "error", errMsg, "duration_ms", duration.Milliseconds())
		return nil, fmt.Errorf("%s", errMsg)
	}

	slog.Debug("exec_completed", "sql", sql, "duration_ms", duration.Milliseconds(), "bytes", stdout.Len())
	return stdout.Bytes(), nil
}

func (e *Executor) RefreshCache(ctx context.Context) error {
	e.mu.Lock()
	e.tables = nil
	e.schemas = make(map[string][]TableColumn)
	e.mu.Unlock()

	_, err := e.listTables(ctx)
	return err
}

func (e *Executor) listTables(ctx context.Context) ([]string, error) {
	e.mu.RLock()
	if len(e.tables) > 0 {
		tables := append([]string(nil), e.tables...)
		e.mu.RUnlock()
		return tables, nil
	}
	e.mu.RUnlock()

	ctx, cancel := context.WithTimeout(ctx, e.timeout)
	defer cancel()

	slog.Debug("listing_tables")
	cmd := exec.CommandContext(ctx, e.binaryPath, "--config_path=/dev/null")
	cmd.Stdin = strings.NewReader(".tables\n")
	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	start := time.Now()
	err := cmd.Run()
	duration := time.Since(start)

	if err != nil {
		errMsg := strings.TrimSpace(stderr.String())
		if errMsg == "" {
			slog.Debug("exec_failed", "op", "list_tables", "error", err, "duration_ms", duration.Milliseconds())
			return nil, err
		}
		slog.Debug("exec_failed", "op", "list_tables", "error", errMsg, "duration_ms", duration.Milliseconds())
		return nil, fmt.Errorf("%s", errMsg)
	}

	slog.Debug("exec_completed", "op", "list_tables", "duration_ms", duration.Milliseconds())

	// osqueryi .tables output looks like:
	//   => table_name
	//   => other_table
	lines := strings.Split(stdout.String(), "\n")
	var tables []string
	for _, line := range lines {
		line = strings.TrimSpace(line)
		if strings.HasPrefix(line, "=> ") {
			tables = append(tables, strings.TrimPrefix(line, "=> "))
		}
	}

	sort.Strings(tables)

	e.mu.Lock()
	if len(e.tables) == 0 {
		e.tables = append([]string(nil), tables...)
	}
	e.mu.Unlock()

	return tables, nil
}

var tableNameRegex = regexp.MustCompile(`^[a-z][a-z0-9_]*$`)

func (e *Executor) validateTableName(name string) error {
	if !tableNameRegex.MatchString(name) {
		return fmt.Errorf("invalid table name: %s", name)
	}
	return nil
}

func (e *Executor) ensureKnownTable(ctx context.Context, tableName string) error {
	if err := e.validateTableName(tableName); err != nil {
		return err
	}

	tables, err := e.listTables(ctx)
	if err != nil {
		return err
	}

	for _, table := range tables {
		if table == tableName {
			return nil
		}
	}

	return fmt.Errorf("unknown table: %s", tableName)
}

func (e *Executor) describeTableColumns(ctx context.Context, tableName string) ([]TableColumn, error) {
	if err := e.ensureKnownTable(ctx, tableName); err != nil {
		return nil, err
	}

	e.mu.RLock()
	if columns, ok := e.schemas[tableName]; ok {
		cached := append([]TableColumn(nil), columns...)
		e.mu.RUnlock()
		return cached, nil
	}
	e.mu.RUnlock()

	query := fmt.Sprintf("PRAGMA table_info(%s);", tableName)
	data, err := e.runSQL(ctx, query)
	if err != nil {
		return nil, err
	}

	var columns []TableColumn
	if err := json.Unmarshal(data, &columns); err != nil {
		return nil, fmt.Errorf("failed to parse schema for %s: %w", tableName, err)
	}

	e.mu.Lock()
	if _, ok := e.schemas[tableName]; !ok {
		e.schemas[tableName] = append([]TableColumn(nil), columns...)
	}
	e.mu.Unlock()

	return append([]TableColumn(nil), columns...), nil
}

func normalizeLimit(limit, defaultValue, maxValue int) int {
	if limit <= 0 {
		return defaultValue
	}
	if limit > maxValue {
		return maxValue
	}
	return limit
}

func trimOptionalSemicolon(value string) string {
	return strings.TrimSuffix(strings.TrimSpace(value), ";")
}

func columnSet(columns []TableColumn) map[string]struct{} {
	valid := make(map[string]struct{}, len(columns))
	for _, column := range columns {
		valid[column.Name] = struct{}{}
	}
	return valid
}

func validateSelectedColumns(selected []string, valid map[string]struct{}) ([]string, error) {
	if len(selected) == 0 {
		return nil, nil
	}

	normalized := make([]string, 0, len(selected))
	for _, column := range selected {
		column = strings.TrimSpace(column)
		if column == "" {
			continue
		}
		if _, ok := valid[column]; !ok {
			return nil, fmt.Errorf("unknown column: %s", column)
		}
		normalized = append(normalized, column)
	}

	if len(normalized) == 0 {
		return nil, nil
	}

	return normalized, nil
}

func validateOrderBy(orderBy []string, valid map[string]struct{}) ([]string, error) {
	normalized := make([]string, 0, len(orderBy))

	for _, raw := range orderBy {
		raw = trimOptionalSemicolon(raw)
		if raw == "" {
			continue
		}

		parts := strings.Fields(raw)
		if len(parts) == 0 || len(parts) > 2 {
			return nil, fmt.Errorf("invalid order_by clause: %s", raw)
		}

		column := parts[0]
		if _, ok := valid[column]; !ok {
			return nil, fmt.Errorf("unknown order_by column: %s", column)
		}

		if len(parts) == 1 {
			normalized = append(normalized, column)
			continue
		}

		direction := strings.ToUpper(parts[1])
		if direction != "ASC" && direction != "DESC" {
			return nil, fmt.Errorf("invalid order_by direction in clause: %s", raw)
		}

		normalized = append(normalized, column+" "+direction)
	}

	return normalized, nil
}

func validateWhereClause(where string) (string, error) {
	where = trimOptionalSemicolon(where)
	if where == "" {
		return "", nil
	}
	if strings.Contains(where, ";") {
		return "", fmt.Errorf("where clause must not contain semicolons")
	}
	return where, nil
}

func (e *Executor) DescribeTable(ctx context.Context, tableName string) ([]byte, error) {
	columns, err := e.describeTableColumns(ctx, tableName)
	if err != nil {
		return nil, err
	}
	return json.Marshal(columns)
}

func (e *Executor) RunQuery(ctx context.Context, sql string) ([]byte, error) {
	return e.runSQL(ctx, sql)
}

func (e *Executor) SearchTables(ctx context.Context, query string, searchColumns bool, limit int) ([]byte, error) {
	query = strings.ToLower(strings.TrimSpace(query))
	if query == "" {
		return nil, fmt.Errorf("missing search query")
	}

	limit = normalizeLimit(limit, 20, 100)

	tables, err := e.listTables(ctx)
	if err != nil {
		return nil, err
	}

	matches := make([]SearchMatch, 0)
	for _, table := range tables {
		match := SearchMatch{TableName: table}
		if strings.Contains(strings.ToLower(table), query) {
			match.MatchReasons = append(match.MatchReasons, "table_name")
		}

		if searchColumns {
			columns, err := e.describeTableColumns(ctx, table)
			if err != nil {
				return nil, err
			}
			for _, column := range columns {
				if strings.Contains(strings.ToLower(column.Name), query) {
					match.MatchingColumns = append(match.MatchingColumns, column.Name)
				}
			}
			if len(match.MatchingColumns) > 0 {
				match.MatchReasons = append(match.MatchReasons, "columns")
			}
		}

		if len(match.MatchReasons) > 0 {
			matches = append(matches, match)
		}
	}

	sort.Slice(matches, func(i, j int) bool {
		iTable := slicesContains(matches[i].MatchReasons, "table_name")
		jTable := slicesContains(matches[j].MatchReasons, "table_name")
		if iTable != jTable {
			return iTable
		}
		if len(matches[i].MatchingColumns) != len(matches[j].MatchingColumns) {
			return len(matches[i].MatchingColumns) > len(matches[j].MatchingColumns)
		}
		return matches[i].TableName < matches[j].TableName
	})

	if len(matches) > limit {
		matches = matches[:limit]
	}

	return json.Marshal(matches)
}

func slicesContains(values []string, target string) bool {
	for _, value := range values {
		if value == target {
			return true
		}
	}
	return false
}

func (e *Executor) PreviewTable(ctx context.Context, tableName string, limit int) ([]byte, error) {
	columns, err := e.describeTableColumns(ctx, tableName)
	if err != nil {
		return nil, err
	}

	limit = normalizeLimit(limit, 5, 100)
	query := fmt.Sprintf("SELECT * FROM %s LIMIT %d;", tableName, limit)
	rows, err := e.runSQL(ctx, query)
	if err != nil {
		return nil, err
	}

	var previewRows []map[string]any
	if err := json.Unmarshal(rows, &previewRows); err != nil {
		return nil, fmt.Errorf("failed to parse preview rows for %s: %w", tableName, err)
	}

	preview := struct {
		TableName string           `json:"table_name"`
		Columns   []TableColumn    `json:"columns"`
		Rows      []map[string]any `json:"rows"`
	}{
		TableName: tableName,
		Columns:   columns,
		Rows:      previewRows,
	}

	return json.Marshal(preview)
}

func (e *Executor) QueryTable(ctx context.Context, tableName string, columns []string, where string, orderBy []string, limit int) ([]byte, error) {
	schema, err := e.describeTableColumns(ctx, tableName)
	if err != nil {
		return nil, err
	}

	validColumns := columnSet(schema)
	columns, err = validateSelectedColumns(columns, validColumns)
	if err != nil {
		return nil, err
	}

	where, err = validateWhereClause(where)
	if err != nil {
		return nil, err
	}

	orderBy, err = validateOrderBy(orderBy, validColumns)
	if err != nil {
		return nil, err
	}

	limit = normalizeLimit(limit, 50, 1000)

	selectedColumns := "*"
	if len(columns) > 0 {
		selectedColumns = strings.Join(columns, ", ")
	}

	var builder strings.Builder
	builder.WriteString("SELECT ")
	builder.WriteString(selectedColumns)
	builder.WriteString(" FROM ")
	builder.WriteString(tableName)

	if where != "" {
		builder.WriteString(" WHERE ")
		builder.WriteString(where)
	}

	if len(orderBy) > 0 {
		builder.WriteString(" ORDER BY ")
		builder.WriteString(strings.Join(orderBy, ", "))
	}

	builder.WriteString(fmt.Sprintf(" LIMIT %d;", limit))

	return e.runSQL(ctx, builder.String())
}

File: rust/src/executor.rs
use regex::Regex;
use serde::{Deserialize, Serialize};
use std::collections::{HashMap, HashSet};
use std::time::Duration;
use tokio::fs;
use tokio::process::Command;
use tokio::sync::{Mutex, RwLock};
use tokio::time::timeout;

const MAX_PAYLOAD_SIZE: usize = 16384; // 16KB soft limit
const MAX_ROWS: usize = 100;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct TableColumn {
    pub cid: String,
    pub dflt_value: String,
    pub name: String,
    pub notnull: String,
    pub pk: String,
    pub r#type: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PersistedCache {
    pub tables: Vec<String>,
    pub schemas: HashMap<String, Vec<TableColumn>>,
    pub all_schemas_fetched: bool,
}

pub struct ExecutorState {
    pub tables: Vec<String>,
    pub schemas: HashMap<String, Vec<TableColumn>>,
    pub all_schemas_fetched: bool,
}

pub struct Executor {
    binary_path: String,
    timeout: Duration,
    cache_file: String,
    state: RwLock<ExecutorState>,
    warm_mu: Mutex<()>,
    table_name_regex: Regex,
}

#[derive(Deserialize)]
struct TableListEntry {
    name: String,
}

#[derive(Deserialize)]
struct FullSchemaEntry {
    #[serde(flatten)]
    column: TableColumn,
    table_name: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SearchMatch {
    pub table_name: String,
    pub match_reasons: Vec<String>,
    #[serde(skip_serializing_if = "Vec::is_empty", default)]
    pub matching_columns: Vec<String>,
}

#[derive(Serialize)]
pub struct QueryResult {
    pub truncated: bool,
    #[serde(skip_serializing_if = "String::is_empty", default)]
    pub message: String,
    pub results: serde_json::Value,
}

#[derive(Serialize)]
struct PreviewResult {
    table_name: String,
    columns: Vec<TableColumn>,
    rows: Vec<serde_json::Value>,
    truncated: bool,
}

fn normalize_limit(limit: i32, default_value: i32, max_value: i32) -> i32 {
    if limit <= 0 {
        default_value
    } else if limit > max_value {
        max_value
    } else {
        limit
    }
}

fn trim_optional_semicolon(value: &str) -> &str {
    value.trim().trim_end_matches(';')
}

fn column_set(columns: &[TableColumn]) -> HashSet<String> {
    columns.iter().map(|col| col.name.clone()).collect()
}

fn validate_selected_columns(
    selected: &[String],
    valid: &HashSet<String>,
) -> Result<Vec<String>, String> {
    if selected.is_empty() {
        return Ok(Vec::new());
    }
    let mut normalized = Vec::new();
    for col in selected {
        let trimmed = col.trim();
        if trimmed.is_empty() {
            continue;
        }
        if !valid.contains(trimmed) {
            return Err(format!("unknown column: {}", trimmed));
        }
        normalized.push(trimmed.to_string());
    }
    Ok(normalized)
}

fn validate_order_by(order_by: &[String], valid: &HashSet<String>) -> Result<Vec<String>, String> {
    let mut normalized = Vec::new();
    for raw in order_by {
        let trimmed = trim_optional_semicolon(raw);
        if trimmed.is_empty() {
            continue;
        }
        let parts: Vec<&str> = trimmed.split_whitespace().collect();
        if parts.is_empty() || parts.len() > 2 {
            return Err(format!("invalid order_by clause: {}", raw));
        }
        let column = parts[0];
        if !valid.contains(column) {
            return Err(format!("unknown order_by column: {}", column));
        }
        if parts.len() == 1 {
            normalized.push(column.to_string());
            continue;
        }
        let direction = parts[1].to_uppercase();
        if direction != "ASC" && direction != "DESC" {
            return Err(format!("invalid order_by direction in clause: {}", raw));
        }
        normalized.push(format!("{} {}", column, direction));
    }
    Ok(normalized)
}

fn validate_where_clause(where_clause: &str) -> Result<String, String> {
    let trimmed = trim_optional_semicolon(where_clause);
    if trimmed.is_empty() {
        return Ok(String::new());
    }
    if trimmed.contains(';') {
        return Err("where clause must not contain semicolons".to_string());
    }
    Ok(trimmed.to_string())
}

fn truncate_json(data: &[u8]) -> Result<Vec<u8>, String> {
    if data.len() < MAX_PAYLOAD_SIZE {
        let mut results: Vec<serde_json::Value> = serde_json::from_slice(data)
            .map_err(|e| format!("failed to parse results for wrapping: {}", e))?;

        let mut truncated = false;
        let mut msg = String::new();
        if results.len() > MAX_ROWS {
            results.truncate(MAX_ROWS);
            truncated = true;
            msg = format!(
                "Results truncated to {} rows to stay within limits.",
                MAX_ROWS
            );
        }

        let res = QueryResult {
            truncated,
            message: msg,
            results: serde_json::Value::Array(results),
        };
        return serde_json::to_vec(&res).map_err(|e| e.to_string());
    }

    let mut results: Vec<serde_json::Value> = serde_json::from_slice(data)
        .map_err(|e| format!("failed to parse results for truncation: {}", e))?;

    let original_count = results.len();
    if original_count > MAX_ROWS {
        results.truncate(MAX_ROWS);
    }

    while results.len() > 1 {
        let new_data = serde_json::to_vec(&results).map_err(|e| e.to_string())?;
        if new_data.len() < MAX_PAYLOAD_SIZE {
            break;
        }
        let half = results.len() / 2;
        results.truncate(half);
    }

    let res = QueryResult {
        truncated: true,
        message: format!(
            "Results truncated from {} to {} rows to stay within size limits.",
            original_count,
            results.len()
        ),
        results: serde_json::Value::Array(results),
    };
    serde_json::to_vec(&res).map_err(|e| e.to_string())
}

impl Executor {
    pub fn new(binary_path: &str, timeout: Duration, cache_file: &str) -> Self {
        let regex = Regex::new(r"^[a-z][a-z0-9_]*$").unwrap();
        let state = ExecutorState {
            tables: Vec::new(),
            schemas: HashMap::new(),
            all_schemas_fetched: false,
        };
        let e = Self {
            binary_path: binary_path.to_string(),
            timeout,
            cache_file: cache_file.to_string(),
            state: RwLock::new(state),
            warm_mu: Mutex::new(()),
            table_name_regex: regex,
        };

        if !cache_file.is_empty() && cache_file != "off" {
            if let Ok(data) = std::fs::read(cache_file) {
                if let Ok(pc) = serde_json::from_slice::<PersistedCache>(&data) {
                    let mut st = e.state.try_write().unwrap();
                    st.tables = pc.tables;
                    st.schemas = pc.schemas;
                    st.all_schemas_fetched = pc.all_schemas_fetched;
                    tracing::info!(file = %cache_file, tables = st.tables.len(), "loaded cache from disk");
                } else {
                    tracing::warn!(file = %cache_file, "failed to parse cache JSON on startup");
                }
            } else {
                tracing::info!(file = %cache_file, "cache file not found, will create on exit or after warming");
            }
        }
        e
    }

    pub async fn save_cache(&self) -> Result<(), String> {
        if self.cache_file.is_empty() || self.cache_file == "off" {
            return Ok(());
        }

        tracing::info!(file = %self.cache_file, "saving cache to disk");

        let pc = {
            let state = self.state.read().await;
            PersistedCache {
                tables: state.tables.clone(),
                schemas: state.schemas.clone(),
                all_schemas_fetched: state.all_schemas_fetched,
            }
        };

        let data = serde_json::to_vec_pretty(&pc)
            .map_err(|e| format!("failed to marshal cache: {}", e))?;

        fs::write(&self.cache_file, data).await.map_err(|e| {
            tracing::error!(file = %self.cache_file, error = %e, "failed to write cache file");
            e.to_string()
        })?;

        tracing::info!(file = %self.cache_file, tables = pc.tables.len(), "saved cache to disk");
        Ok(())
    }

    pub async fn run_sql(&self, sql: &str) -> Result<Vec<u8>, String> {
        tracing::debug!(sql = sql, "executing_sql");

        let mut cmd = Command::new(&self.binary_path);
        cmd.args(&["--json", "--config_path=/dev/null", sql]);
        cmd.kill_on_drop(true);

        let start = std::time::Instant::now();
        let output_fut = cmd.output();

        match timeout(self.timeout, output_fut).await {
            Ok(Ok(output)) => {
                let duration_ms = start.elapsed().as_millis();
                if output.status.success() {
                    tracing::debug!(
                        sql = sql,
                        duration_ms = duration_ms,
                        bytes = output.stdout.len(),
                        "exec_completed"
                    );
                    Ok(output.stdout)
                } else {
                    let err_msg = String::from_utf8_lossy(&output.stderr).trim().to_string();
                    tracing::debug!(sql = sql, error = %err_msg, duration_ms = duration_ms, "exec_failed");
                    Err(if err_msg.is_empty() {
                        format!("osqueryi exited with status {}", output.status)
                    } else {
                        err_msg
                    })
                }
            }
            Ok(Err(e)) => {
                let duration_ms = start.elapsed().as_millis();
                tracing::debug!(sql = sql, error = %e, duration_ms = duration_ms, "exec_failed");
                Err(e.to_string())
            }
            Err(_) => {
                let duration_ms = start.elapsed().as_millis();
                tracing::debug!(
                    sql = sql,
                    error = "timeout",
                    duration_ms = duration_ms,
                    "exec_failed"
                );
                Err("query timeout exceeded".to_string())
            }
        }
    }

    pub async fn list_tables(&self) -> Result<Vec<String>, String> {
        {
            let state = self.state.read().await;
            if !state.tables.is_empty() {
                return Ok(state.tables.clone());
            }
        }

        tracing::debug!("listing_tables");
        let data = self
            .run_sql("SELECT name FROM osquery_registry WHERE registry = 'table' ORDER BY name;")
            .await?;
        let results: Vec<TableListEntry> = serde_json::from_slice(&data)
            .map_err(|e| format!("failed to parse table list: {}", e))?;

        let tables: Vec<String> = results.into_iter().map(|entry| entry.name).collect();

        let mut state = self.state.write().await;
        if state.tables.is_empty() {
            state.tables = tables.clone();
            drop(state);
            let _ = self.save_cache().await;
        }

        Ok(tables)
    }

    pub async fn fetch_all_schemas(&self) -> Result<(), String> {
        tracing::debug!("fetching_all_schemas");
        let query = "SELECT m.name as table_name, p.* FROM (SELECT name FROM osquery_registry WHERE registry = 'table') m, pragma_table_info(m.name) p;";
        let data = self.run_sql(query).await?;

        let entries: Vec<FullSchemaEntry> = serde_json::from_slice(&data)
            .map_err(|e| format!("failed to parse all schemas: {}", e))?;

        let mut new_schemas: HashMap<String, Vec<TableColumn>> = HashMap::new();
        for entry in entries {
            new_schemas
                .entry(entry.table_name)
                .or_default()
                .push(entry.column);
        }

        let mut state = self.state.write().await;
        for (table, cols) in new_schemas {
            state.schemas.insert(table, cols);
        }

        if state.tables.is_empty() {
            let mut tables: Vec<String> = state.schemas.keys().cloned().collect();
            tables.sort();
            state.tables = tables;
        }
        state.all_schemas_fetched = true;
        drop(state);

        let _ = self.save_cache().await;
        Ok(())
    }

    pub async fn ensure_all_schemas(&self) -> Result<(), String> {
        {
            let state = self.state.read().await;
            if state.all_schemas_fetched {
                return Ok(());
            }
        }

        let _guard = self.warm_mu.lock().await;
        {
            let state = self.state.read().await;
            if state.all_schemas_fetched {
                return Ok(());
            }
        }

        self.fetch_all_schemas().await
    }

    pub async fn refresh_cache(&self) -> Result<(), String> {
        let _guard = self.warm_mu.lock().await;

        {
            let mut state = self.state.write().await;
            state.tables.clear();
            state.schemas.clear();
            state.all_schemas_fetched = false;
        }

        self.fetch_all_schemas().await
    }

    pub fn validate_table_name(&self, name: &str) -> Result<(), String> {
        if !self.table_name_regex.is_match(name) {
            return Err(format!("invalid table name: {}", name));
        }
        Ok(())
    }

    pub async fn ensure_known_table(&self, table_name: &str) -> Result<(), String> {
        self.validate_table_name(table_name)?;
        let tables = self.list_tables().await?;
        if tables.iter().any(|t| t == table_name) {
            Ok(())
        } else {
            Err(format!("unknown table: {}", table_name))
        }
    }

    pub async fn describe_table_columns(
        &self,
        table_name: &str,
    ) -> Result<Vec<TableColumn>, String> {
        self.ensure_known_table(table_name).await?;

        {
            let state = self.state.read().await;
            if let Some(columns) = state.schemas.get(table_name) {
                return Ok(columns.clone());
            }
        }

        let query = format!("PRAGMA table_info({});", table_name);
        let data = self.run_sql(&query).await?;
        let columns: Vec<TableColumn> = serde_json::from_slice(&data)
            .map_err(|e| format!("failed to parse schema for {}: {}", table_name, e))?;

        let mut state = self.state.write().await;
        if !state.schemas.contains_key(table_name) {
            state
                .schemas
                .insert(table_name.to_string(), columns.clone());
            drop(state);
            let _ = self.save_cache().await;
        }

        Ok(columns)
    }

    pub async fn describe_table(&self, table_name: &str) -> Result<Vec<u8>, String> {
        let columns = self.describe_table_columns(table_name).await?;
        serde_json::to_vec(&columns).map_err(|e| e.to_string())
    }

    pub async fn run_query(&self, sql: &str) -> Result<Vec<u8>, String> {
        let data = self.run_sql(sql).await?;
        truncate_json(&data)
    }

    pub async fn search_tables(
        &self,
        query: &str,
        search_columns: bool,
        limit: i32,
    ) -> Result<Vec<u8>, String> {
        let query_lower = query.to_lowercase().trim().to_string();
        if query_lower.is_empty() {
            return Err("missing search query".to_string());
        }

        let limit_val = normalize_limit(limit, 20, 100) as usize;
        let tables = self.list_tables().await?;

        if search_columns {
            self.ensure_all_schemas().await?;
        }

        let mut matches = Vec::new();

        if search_columns {
            let state = self.state.read().await;
            for table in &tables {
                let mut match_entry = SearchMatch {
                    table_name: table.clone(),
                    match_reasons: Vec::new(),
                    matching_columns: Vec::new(),
                };

                let table_lower = table.to_lowercase();
                if table_lower.contains(&query_lower) {
                    match_entry.match_reasons.push("table_name".to_string());
                }

                if let Some(columns) = state.schemas.get(table) {
                    for col in columns {
                        if col.name.to_lowercase().contains(&query_lower) {
                            match_entry.matching_columns.push(col.name.clone());
                        }
                    }
                }

                if !match_entry.matching_columns.is_empty() {
                    match_entry.match_reasons.push("columns".to_string());
                }

                if !match_entry.match_reasons.is_empty() {
                    matches.push(match_entry);
                }
            }
        } else {
            for table in &tables {
                let table_lower = table.to_lowercase();
                if table_lower.contains(&query_lower) {
                    let match_entry = SearchMatch {
                        table_name: table.clone(),
                        match_reasons: vec!["table_name".to_string()],
                        matching_columns: Vec::new(),
                    };
                    matches.push(match_entry);
                }
            }
        }

        matches.sort_by(|a, b| {
            let a_has_table = a.match_reasons.iter().any(|r| r == "table_name");
            let b_has_table = b.match_reasons.iter().any(|r| r == "table_name");
            if a_has_table != b_has_table {
                return b_has_table.cmp(&a_has_table);
            }
            let a_col_count = a.matching_columns.len();
            let b_col_count = b.matching_columns.len();
            if a_col_count != b_col_count {
                return b_col_count.cmp(&a_col_count);
            }
            a.table_name.cmp(&b.table_name)
        });

        if matches.len() > limit_val {
            matches.truncate(limit_val);
        }

        serde_json::to_vec(&matches).map_err(|e| e.to_string())
    }

    pub async fn preview_table(&self, table_name: &str, limit: i32) -> Result<Vec<u8>, String> {
        let columns = self.describe_table_columns(table_name).await?;
        let limit_val = normalize_limit(limit, 5, 100);

        let query = format!("SELECT * FROM {} LIMIT {};", table_name, limit_val);
        let rows_data = self.run_sql(&query).await?;

        let mut preview_rows: Vec<serde_json::Value> = serde_json::from_slice(&rows_data)
            .map_err(|e| format!("failed to parse preview rows for {}: {}", table_name, e))?;

        let mut truncated = false;
        if preview_rows.len() > MAX_ROWS {
            preview_rows.truncate(MAX_ROWS);
            truncated = true;
        }

        let mut preview = PreviewResult {
            table_name: table_name.to_string(),
            columns,
            rows: preview_rows,
            truncated,
        };

        let mut data = serde_json::to_vec(&preview).map_err(|e| e.to_string())?;

        if data.len() > MAX_PAYLOAD_SIZE {
            while preview.rows.len() > 1 {
                let half = preview.rows.len() / 2;
                preview.rows.truncate(half);
                preview.truncated = true;
                data = serde_json::to_vec(&preview).map_err(|e| e.to_string())?;
                if data.len() < MAX_PAYLOAD_SIZE {
                    break;
                }
            }
        }

        Ok(data)
    }

    pub async fn query_table(
        &self,
        table_name: &str,
        columns: &[String],
        where_clause: &str,
        order_by: &[String],
        limit: i32,
    ) -> Result<Vec<u8>, String> {
        let schema = self.describe_table_columns(table_name).await?;
        let valid_columns = column_set(&schema);

        let validated_columns = validate_selected_columns(columns, &valid_columns)?;
        let validated_where = validate_where_clause(where_clause)?;
        let validated_order_by = validate_order_by(order_by, &valid_columns)?;

        let limit_val = normalize_limit(limit, 50, 1000);

        let selected_columns = if validated_columns.is_empty() {
            "*".to_string()
        } else {
            validated_columns.join(", ")
        };

        let mut sql = format!("SELECT {} FROM {}", selected_columns, table_name);

        if !validated_where.is_empty() {
            sql.push_str(" WHERE ");
            sql.push_str(&validated_where);
        }

        if !validated_order_by.is_empty() {
            sql.push_str(" ORDER BY ");
            sql.push_str(&validated_order_by.join(", "));
        }

        sql.push_str(&format!(" LIMIT {};", limit_val));

        let data = self.run_sql(&sql).await?;
        truncate_json(&data)
    }
}
